### ISSUE TITLE
Required Workflows

### ISSUE THREAD

**Issue by jimafisk** [go-gitea/gitea#26666]
### Feature Description

**Feature Request:** As a Gitea Site Administrator I'd love to be able to universally enforce an Action on all repos that users can't edit/change. 

**Why it's useful:** Opening up CI/CD to users comes with some risk for abuse (see [this example](https://about.gitlab.com/blog/2021/05/17/prevent-crypto-mining-abuse/) of GitLab requiring valid credit cards to avoid crypto mining on their runners). Gitea is also perfectly aligned for folks who want to run niche SaaS products that might use the same tech / workflow across repos.

**Similar (but different) features seen on GitHub:**

GitHub Actions have [Required Workflows](https://docs.github.com/en/actions/using-workflows/required-workflows) that allow you to specify CI/CD operations that run across multiple repositories. See more details in [their blog post](https://github.blog/2023-01-10-introducing-required-workflows-and-configuration-variables-to-github-actions/). My understanding is this is similar to [Reusable Workflows](https://docs.github.com/en/actions/using-workflows/reusing-workflows) (related [blog post](https://github.blog/2022-02-10-using-reusable-workflows-github-actions/)) but differs slightly: 
> Required workflows can be stored in any repository folder and are not restricted to the .github/workflows folder like normal workflows. If a required workflow calls a reusable workflow, the reusable workflow must be stored in the .github/workflows folder.

For my particular use case, I wouldn't necessarily want to define the Action in another repo, I'd rather set it in the Site Administration settings or in a separate config file.

**Comment by lunny** [comment]
How about adding it to protected branch rules?

**Comment by jimafisk** [comment]
Thanks for following up @lunny! I'm not sure if my use case is common, but my hope was the universal action wouldn't actually get tracked with the repo code itself. I pictured being able to define the action in something like `Site Administrator` > `Actions` > `Global Workflows` and being able to assign it to all or select repositories throughout the instance. It would allow administrators to enforce certain repo rules, or in my case our whole Gitea instance is centered around a specific technology and we want to require a particular build/deployment process that users can't change. Maybe what you had in mind for protected branch rules is similar? Thanks!

**Comment by lunny** [comment]
> Thanks for following up @lunny! I'm not sure if my use case is common, but my hope was the universal action wouldn't actually get tracked with the repo code itself. I pictured being able to define the action in something like `Site Administrator` > `Actions` > `Global Workflows` and being able to assign it to all or select repositories throughout the instance. It would allow administrators to enforce certain repo rules, or in my case our whole Gitea instance is centered around a specific technology and we want to require a particular build/deployment process that users can't change. Maybe what you had in mind for protected branch rules is similar? Thanks!

OK. I understand your requirements. It's a global enforce workflows which is configured by admin in admin panel.

**Comment by jimafisk** [comment]
> Whether the logs should be displayed on repository's actions tab or not?

That's a great question, for me it would be useful to still log the builds in the repo. Potentially other folks would want to hide this though - I guess having a configurable "hide logs" option in the administrator settings would be ideal, but not critical.

**Comment by AvengerMoJo** [comment]
@lunny I started with creating a WIP branch https://github.com/AvengerMoJo/gitea/tree/wip_require_action_feature
It allows the repo to enable a global require workflow flag to start with. ( done ) 
Then, I will create a org setting page to adding the global enabled workflow to auto run for a ll repo in the org. (wip)

**Comment by AvengerMoJo** [comment]
![captures_chrome-capture-2024-2-4 (2)](https://github.com/go-gitea/gitea/assets/109793/7671b39e-a3d7-400d-b05a-887d1fda3bd4)
I am following the action runner template and then added a "add global workflow" button so, all the enabled workflows are listed in the pull down, then select / multi select to enable for all the repos in org to be called.  
I may need some help and discussion before implementing next step. 
Once the global enable workflow (Require Action) is being added. 
1) Default behavior is to allow all repos in the Org to run those Require Actions. Do we need the filter function now to remove specific repo from running it? 
2) How should I schedule those Require Actions schedule to run for each repo? 
3) Since each repo doesn't has actually has the workflow files in their git tree. The original Actions tab in the repo interface need to be updated, but where should those log and Require Actions history should be store? 

Thanks

**Comment by lunny** [comment]
> ![captures_chrome-capture-2024-2-4 (2)](https://private-user-images.githubusercontent.com/109793/309682656-7671b39e-a3d7-400d-b05a-887d1fda3bd4.jpeg?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDk1NDI5OTgsIm5iZiI6MTcwOTU0MjY5OCwicGF0aCI6Ii8xMDk3OTMvMzA5NjgyNjU2LTc2NzFiMzllLWEzZDctNDAwZC1iMDVhLTg4N2QxZmRhM2JkNC5qcGVnP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDMwNCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAzMDRUMDg1ODE4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZmEwOThhNTQ4NTA3MjU5NGI2OWJiNzA3NWVlMjczM2VjNDhiMGMwNzM4ZWM0Y2FmZjRjNmQwMGRhMmRkOTk2MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.I8dPasMza4egarYEHs0oTOWq7ODxLiOpTU2oU9P0OOs) I am following the action runner template and then added a "add global workflow" button so, all the enabled workflows are listed in the pull down, then select / multi select to enable for all the repos in org to be called. I may need some help and discussion before implementing next step. Once the global enable workflow (Require Action) is being added.
> 
>     1. Default behavior is to allow all repos in the Org to run those Require Actions. Do we need the filter function now to remove specific repo from running it?

Maybe the repositories could choose whether to inherit from these global workflows.

>     2. How should I schedule those Require Actions schedule to run for each repo?

Hm, the get schedules function needs to be changed to get parent-level workflows.

>     3. Since each repo doesn't has actually has the workflow files in their git tree. The original Actions tab in the repo interface need to be updated, but where should those log and Require Actions history should be store?

The inherited workflow files should be displayed on actions left sidebar with different color/icon with repo's. Logs should be still stored like before. I have no idea how to store the inherited workflows.

**Comment by AvengerMoJo** [comment]
> > ```
> > 1. Default behavior is to allow all repos in the Org to run those Require Actions. Do we need the filter function now to remove specific repo from running it?
> > ```
> 
> Maybe the repositories could choose whether to inherit from these global workflows.
> 
I don't object to the idea of repo level should able to opt out of the global workflows. However, I want to highlight and make sure we are in-sync of the benefit about this Require Action is to allow Org level to apply workflow to all Repos without adding files or variable and setup them up one by one individually. So the Org level view should be the top level default to enable for all repos. ( Or else individual repo could actually use or link other workflow privately not need for this feature ) and likewise Org level filter may need to implement to opt out the repo don't need to run the global workflows?

**Comment by lunny** [comment]
>Note: GitHub no longer supports required workflows for GitHub Actions. To require workflows to pass before merging, upgrade your GitHub Enterprise Server instance to the latest version and use repository rulesets instead.

Looks like Github has deprecated this feature.

**Comment by AvengerMoJo** [comment]
Here is the current implementation so far. 
![EnableGlobal_Step1](https://github.com/go-gitea/gitea/assets/109793/42ce659e-6ce6-4c30-bf91-e006883b6b1e)
![EnableGlobal_Step2](https://github.com/go-gitea/gitea/assets/109793/e857fd9c-0ea1-491f-8ce8-1adc545b4570)
![Org_Setting_Action_Step1](https://github.com/go-gitea/gitea/assets/109793/2a2e5a0f-763f-45d6-85c2-235aa9bd1216)
![Org_Setting_Action_Add_Step2](https://github.com/go-gitea/gitea/assets/109793/554e7910-0e7c-46a6-9bf4-8d0c8635e94e)
![Org_Setting_Action_Add_Step3](https://github.com/go-gitea/gitea/assets/109793/1ded983c-e327-4de2-b245-66fa6c62c9d5)

**Comment by BlenderDefender** [comment]
@lunny Yes, GitHub has moved this feature to branch protection rules, and it's only available in GitHub enterprise cloud.

**Comment by AvengerMoJo** [comment]
@lunny @BlenderDefender I also notice that, but does it mean gitea action don't want this feature?

**Comment by AvengerMoJo** [comment]
@lunny @jimafisk  I think a pool man wip implementation is ready for you to test. https://github.com/go-gitea/gitea/compare/main...AvengerMoJo:gitea:wip_require_action

**Comment by jimafisk** [comment]
Hi @AvengerMoJo,

Thanks so much for working on this issue! I'm very excited for this :)

I'm trying to test it out, but when I try to build (`TAGS="bindata sqlite sqlite_unlock_notify" make build`) on your [wip_require_action](https://github.com/AvengerMoJo/gitea/tree/wip_require_action) branch it seems to throw these errors:

```
CGO_CFLAGS="-O2 -g -DSQLITE_MAX_VARIABLE_NUMBER=32766" go build -v  -tags 'bindata sqlite sqlite_unlock_notify' -ldflags '-s -w  -X "main.MakeVersion=GNU Make 4.3" -X "main.Version=a8b8608fd" -X "main.Tags=bindata sqlite sqlite_unlock_notify"' -o gitea
code.gitea.io/gitea/routers/web/shared/actions
# code.gitea.io/gitea/routers/web/shared/actions
routers/web/shared/actions/require_action.go:75:12: ctx.ParamsInt64 undefined (type *"code.gitea.io/gitea/services/context".Context has no field or method ParamsInt64)
make: *** [Makefile:797: gitea] Error 1
```

Is there a step I need to be doing to get this to work? Thank you!!

**Comment by AvengerMoJo** [comment]
@jimafisk thanks for testing. I don't know why the patch is missing. I thought I included that. But I push the patch again and please try again. Thanks

**Comment by AvengerMoJo** [comment]
@jimafisk @lunny FYI I'm in the process of writing a document for the require action workflows.  Please let me know how can I start the discussion: I get the following content so far. 

- Introduction
- Feature Description
- How to Use

**Comment by AvengerMoJo** [comment]
@jimafisk @lunny FYI my general documentation: [Global Require Action CI_CD implementation for Gitea.pdf](https://github.com/user-attachments/files/16578441/Global.Require.Action.CI_CD.implementation.for.Gitea.pdf)

**Comment by jimafisk** [comment]
Hi @AvengerMoJo,

Thanks for re-adding the patch and pushing, I'm able to build successfully now! :rocket: 

Is there anything else I need to do to enable this feature? I am using a newer version of Gitea that has actions enabled by default, but I added the following to `custom/conf/app.ini` just in case:

```
[actions]
ENABLE=true
```

 I do see the option to "Enable Global Require" under the "Actions" tab of my Org's repo. If I select this, I do see a red tag that says "Global Enabled" appear under the workflow that I enabled this for. However, under the Org's `Settings` > `Actions` I only see 3 options (`Runners`, `Secrets`, and `Variables`) and do _not_ see the `Require Action` option from the screenshot in your guide PDF (thank you for that!). Similarly, if I go to `Site Administration` > `Actions` I only see 2 options (`Runners` and `Variables`). My apologies if I'm missing a step, but would love to continue testing if you could point me in the right direction! Thank you!

**Comment by AvengerMoJo** [comment]
@jimafisk I don't think you need any thing else. Maybe just the frontend template is not being build completely? Since that extra menu should show no matter. Are you using the following branch? 
https://github.com/AvengerMoJo/gitea/tree/wip_require_action

You can also check 
templates/org/settings/actions.tmpl
there should be a new require_action org/setting/layout_head 

![Gitea-Git-with-a-cup-of-tea](https://github.com/user-attachments/assets/faf8f311-af39-4e75-a60a-eb2ab207ffc5)

**Comment by jimafisk** [comment]
Ho @AvengerMoJo,

I do see the following section in `templates/org/settings/actions.tmpl`:
```go
{{if eq .PageType "require_action"}}
	{{template "shared/actions/require_action_list" .}}
{{else if eq .PageType "runners"}}
```

I also looked at `templates/org/settings/navbar.tmpl` and see:
```html
<a class="{{if .PageIsSharedSettingsRequireAction}}active {{end}}item" href="{{.OrgLink}}/settings/actions/require_action">
	{{ctx.Locale.Tr "actions.require_action"}}
</a>
```

However, even if I edit this file and rebuild, I can't seem to change the output of the links in the left sidebar, so I'm not sure if something is getting cached or what is going on there. I'll keep tinkering to see if I can figure out what's happening. Thanks!

**Comment by AvengerMoJo** [comment]
@jimafisk  :) I don't recall there is any cache around the frontend but did you run 
`TAGS="bindata" make watch-frontend`
